test/integration/change-streams/change_stream.test.ts

```diff
@@ -948,31 +953,178 @@ describe('Change Streams', function () {
       'This test only worked because of timing, changeStream.close does not remove the change listener';
   });
 
-  describe('#tryNext()', function () {
-    it('should return null on single iteration of empty cursor', {
-      metadata: { requires: { topology: 'replicaset' } },
-      async test() {
-        const doc = await changeStream.tryNext();
-        expect(doc).to.be.null;
-      }
+  describe('iterator api', function () {
+    describe('#tryNext()', function () {
+      it('should return null on single iteration of empty cursor', {
+        metadata: { requires: { topology: 'replicaset' } },
+        async test() {
+          const doc = await changeStream.tryNext();
+          expect(doc).to.be.null;
+        }
+      });
+
+      it('should iterate a change stream until first empty batch', {
+        metadata: { requires: { topology: 'replicaset' } },
+        async test() {
+          // tryNext doesn't send the initial agg, just checks the driver document batch cache
+          const firstTry = await changeStream.tryNext();
+          expect(firstTry).to.be.null;
+
+          await initIteratorMode(changeStream);
+          await collection.insertOne({ a: 42 });
+
+          const secondTry = await changeStream.tryNext();
+          expect(secondTry).to.be.an('object');
+
+          const thirdTry = await changeStream.tryNext();
+          expect(thirdTry).to.be.null;
+        }
+      });
     });
 
-    it('should iterate a change stream until first empty batch', {
-      metadata: { requires: { topology: 'replicaset' } },
-      async test() {
-        // tryNext doesn't send the initial agg, just checks the driver document batch cache
-        const firstTry = await changeStream.tryNext();
-        expect(firstTry).to.be.null;
+    describe('#asyncIterator', function () {
+      it(
+        'can iterate through changes',
+        { requires: { topology: '!single', mongodb: '>=4.2' } },
+        async function () {
+          changeStream = collection.watch([]);
+          await initIteratorMode(changeStream);
+          const changeStreamIterator = changeStream[Symbol.asyncIterator]();
```

Comment:
This seems like an odd way of testing the iterator; the goal in the ticket says "Making the ChangeStream an async iterable would allow idiomatic use in for-await loops." Instead of manually calling the next function of the iterator, we should be testing this the way it's intended to be used by our users,
```
for (const change of await changeStream) {
   const { fullDocument } = change.value;
   const expectedDoc = docs.shift();
   expect(fullDocument.city).to.equal(expectedDoc.city);
   if (docs.length === 0) {
     break;
   }
}
expect(docs).to.have.length(0, 'expected to find all docs before exiting loop');
// ^ just a sanity check to make sure we didn't early-exit without error
```

and this goes for all iterator test cases in this file

Reply:
@dariakp Andy wrote the tests this way based on input from me. Andy originally had the tests like you're suggesting but I suggested this approach.

The async iterator method returns a generator.  A for-await loop is just syntactic sugar over manual iteration. I think it's preferable to describe the expected behavior of our generator in the tests by testing each method. The for-await loop will work automatically. 

This also cleans up tests too. Some of the error tests had additional, impossible conditions that we needed to consider when using a for-await loop. For example, consider the test that confirms that when an error is thrown internally, the error is thrown. Originally, we used a for-await loop but the test had an empty loop body (because the first iteration would throw an error and we would never enter the loop), so the options were to either lint the warning away or add an expect.fail.

This approach is 
- More granular
- easier to work with in many cases and requires no breaking from loops
- Completely equivalent to testing using for-await loops

Reply:
What you are describing makes a case for unit testing the behavior of `next()` in the new iterator, and I am certainly not going to object to unit tests being added for this purpose (or these tests being moved or copied to the unit test directory with some adjustments). However, we are talking about integration tests, not unit tests, and we do need to have integration tests that cover *actual use cases* of the feature being implemented. Iterating for-of, and, in particular, breaking out of for-of loops will have very specific behaviors depending on how the iterator is implemented. If we decide to manually create the async iterator or iterable in the future, our integration tests should still be valid. Same if Node.js decides to change exactly how the for-of uses the next() function, or introduces new iteration protocols that we want to use instead of this one. The bottom line is that the feature isn't whether `Symbol.asyncIterator` is defined, it's that the for await of loop works.

References:
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of#iterating_over_sync_iterables_and_generators
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols
- NODE-4598

Reply:
Also, just wanted to add that `expect.fail` actually would be the appropriate thing for an integration test in the error case, because, again, we are testing what our users would expect to see in such a case, and it informs how we document suggested error handling to our users

Reply:
The feature isn't only about for-await loops.  Async generators can be used in other ways:

- https://github.com/ReactiveX/IxJS, a popular functional programming library can now be used with change streams
- change streams can now be used as streams natively with Readable.from

The ticket you linked is one example (for-await) but the actual feature is the async generator we now support.

as far as the tests on this PR go, if you'd like tests using a for-await loop we can add them but I think there's value in testing each part of the async generator individually.

Reply:
From slack:
- We'll use for..await for the main integration test cases (full iteration without errors and with resumable errors; manual break from loop and then creating a new loop; and encountering non-resumable errors)
- There is value in explicitly testing the `asyncIterable`, so we can keep tests that provide additional coverage to that end, with anything that calls for mocking to unit tests

Reply:
Implemented!